From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: BuildTools <unconfigured@null.spigotmc.org>
Date: Mon, 22 Jun 2026 18:05:14 -0400
Subject: [PATCH] fixup! paper File Patches


diff --git a/net/minecraft/world/entity/Entity.java b/net/minecraft/world/entity/Entity.java
index 3bf618f1d756bad7755a87803132bd731e7c41be..c34b757ebe63ef8f330f29a4df6879a12f2ec3d9 100644
--- a/net/minecraft/world/entity/Entity.java
+++ b/net/minecraft/world/entity/Entity.java
@@ -152,7 +152,7 @@ import org.jetbrains.annotations.Contract;
 import org.jspecify.annotations.Nullable;
 import org.slf4j.Logger;
 
-public abstract class Entity implements SyncedDataHolder, DebugValueSource, Nameable, ItemOwner, SlotProvider, EntityAccess, ScoreHolder, DataComponentGetter, ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity, ca.spottedleaf.moonrise.patches.entity_tracker.EntityTrackerEntity { // Paper - rewrite chunk system // Paper - optimise entity tracker
+public abstract class Entity implements SyncedDataHolder, DebugValueSource, Nameable, ItemOwner, SlotProvider, EntityAccess, ScoreHolder, DataComponentGetter {
     // CraftBukkit start
     private static final int CURRENT_LEVEL = 2;
     static boolean isLevelAtLeast(ValueInput input, int level) {
@@ -161,17 +161,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
 
     // Paper start - Share random for entities to make them more random
     public static RandomSource SHARED_RANDOM = new RandomRandomSource();
-    // Paper start - replace random
-    private static final class RandomRandomSource extends ca.spottedleaf.moonrise.common.util.ThreadUnsafeRandom {
-        public RandomRandomSource() {
-            this(net.minecraft.world.level.levelgen.RandomSupport.generateUniqueSeed());
-        }
-
-        public RandomRandomSource(long seed) {
-            super(seed);
-        }
-
-        // Paper end - replace random
+    private static final class RandomRandomSource extends java.util.Random implements net.minecraft.world.level.levelgen.BitRandomSource {
         private boolean locked = false;
 
         @Override
@@ -184,7 +174,61 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             }
         }
 
-        // Paper - replace random
+        @Override
+        public RandomSource fork() {
+            return new net.minecraft.world.level.levelgen.LegacyRandomSource(this.nextLong());
+        }
+
+        @Override
+        public net.minecraft.world.level.levelgen.PositionalRandomFactory forkPositional() {
+            return new net.minecraft.world.level.levelgen.LegacyRandomSource.LegacyPositionalRandomFactory(this.nextLong());
+        }
+
+        // these below are added to fix reobf issues that I don't wanna deal with right now
+        @Override
+        public int next(int bits) {
+            return super.next(bits);
+        }
+
+        @Override
+        public int nextInt(int origin, int bound) {
+            return net.minecraft.world.level.levelgen.BitRandomSource.super.nextInt(origin, bound);
+        }
+
+        @Override
+        public long nextLong() {
+            return net.minecraft.world.level.levelgen.BitRandomSource.super.nextLong();
+        }
+
+        @Override
+        public int nextInt() {
+            return net.minecraft.world.level.levelgen.BitRandomSource.super.nextInt();
+        }
+
+        @Override
+        public int nextInt(int bound) {
+            return net.minecraft.world.level.levelgen.BitRandomSource.super.nextInt(bound);
+        }
+
+        @Override
+        public boolean nextBoolean() {
+            return net.minecraft.world.level.levelgen.BitRandomSource.super.nextBoolean();
+        }
+
+        @Override
+        public float nextFloat() {
+            return net.minecraft.world.level.levelgen.BitRandomSource.super.nextFloat();
+        }
+
+        @Override
+        public double nextDouble() {
+            return net.minecraft.world.level.levelgen.BitRandomSource.super.nextDouble();
+        }
+
+        @Override
+        public double nextGaussian() {
+            return super.nextGaussian();
+        }
     }
     // Paper end - Share random for entities to make them more random
     public org.bukkit.event.entity.CreatureSpawnEvent.@Nullable SpawnReason spawnReason; // Paper - Entity#getEntitySpawnReason
@@ -368,15 +412,6 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     private final int despawnTime; // Paper - entity despawn time limit
     public int totalEntityAge; // Paper - age-like counter for all entities
     public final io.papermc.paper.entity.activation.ActivationType activationType = io.papermc.paper.entity.activation.ActivationType.activationTypeFor(this); // Paper - EAR 2/tracking ranges
-    // Paper start - EAR 2
-    public final boolean defaultActivationState;
-    public long activatedTick = Integer.MIN_VALUE;
-    public boolean isTemporarilyActive;
-    public long activatedImmunityTick = Integer.MIN_VALUE;
-
-    public void inactiveTick() {
-    }
-    // Paper end - EAR 2
     // CraftBukkit end
 
     // Paper start
@@ -384,156 +419,6 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         return this.dimensions.makeBoundingBox(x, y, z);
     }
     // Paper end
-    // Paper start - rewrite chunk system
-    private final boolean isHardColliding = this.moonrise$isHardCollidingUncached();
-    private net.minecraft.server.level.FullChunkStatus chunkStatus;
-    private ca.spottedleaf.moonrise.patches.chunk_system.level.chunk.ChunkData chunkData;
-    private int sectionX = Integer.MIN_VALUE;
-    private int sectionY = Integer.MIN_VALUE;
-    private int sectionZ = Integer.MIN_VALUE;
-    private boolean updatingSectionStatus;
-
-    @Override
-    public final boolean moonrise$isHardColliding() {
-        return this.isHardColliding;
-    }
-
-    @Override
-    public final net.minecraft.server.level.FullChunkStatus moonrise$getChunkStatus() {
-        return this.chunkStatus;
-    }
-
-    @Override
-    public final void moonrise$setChunkStatus(final net.minecraft.server.level.FullChunkStatus status) {
-        this.chunkStatus = status;
-    }
-
-    @Override
-    public final ca.spottedleaf.moonrise.patches.chunk_system.level.chunk.ChunkData moonrise$getChunkData() {
-        return this.chunkData;
-    }
-
-    @Override
-    public final void moonrise$setChunkData(final ca.spottedleaf.moonrise.patches.chunk_system.level.chunk.ChunkData chunkData) {
-        this.chunkData = chunkData;
-    }
-
-    @Override
-    public final int moonrise$getSectionX() {
-        return this.sectionX;
-    }
-
-    @Override
-    public final void moonrise$setSectionX(final int x) {
-        this.sectionX = x;
-    }
-
-    @Override
-    public final int moonrise$getSectionY() {
-        return this.sectionY;
-    }
-
-    @Override
-    public final void moonrise$setSectionY(final int y) {
-        this.sectionY = y;
-    }
-
-    @Override
-    public final int moonrise$getSectionZ() {
-        return this.sectionZ;
-    }
-
-    @Override
-    public final void moonrise$setSectionZ(final int z) {
-        this.sectionZ = z;
-    }
-
-    @Override
-    public final boolean moonrise$isUpdatingSectionStatus() {
-        return this.updatingSectionStatus;
-    }
-
-    @Override
-    public final void moonrise$setUpdatingSectionStatus(final boolean to) {
-        this.updatingSectionStatus = to;
-    }
-
-    @Override
-    public final boolean moonrise$hasAnyPlayerPassengers() {
-        if (this.passengers.isEmpty()) {
-            return false;
-        }
-        return this.getIndirectPassengersStream().anyMatch((entity) -> entity instanceof Player);
-    }
-    // Paper end - rewrite chunk system
-    // Paper start - optimise collisions
-    private static float[] calculateStepHeights(final AABB box, final List<VoxelShape> voxels, final List<AABB> aabbs, final float stepHeight,
-                                                final float collidedY) {
-        final FloatArraySet ret = new FloatArraySet();
-
-        for (int i = 0, len = voxels.size(); i < len; ++i) {
-            final VoxelShape shape = voxels.get(i);
-
-            final double[] yCoords = ((ca.spottedleaf.moonrise.patches.collisions.shape.CollisionVoxelShape)shape).moonrise$rootCoordinatesY();
-            final double yOffset = ((ca.spottedleaf.moonrise.patches.collisions.shape.CollisionVoxelShape)shape).moonrise$offsetY();
-
-            for (final double yUnoffset : yCoords) {
-                final double y = yUnoffset + yOffset;
-
-                final float step = (float)(y - box.minY);
-
-                if (step > stepHeight) {
-                    break;
-                }
-
-                if (step < 0.0f || !(step != collidedY)) {
-                    continue;
-                }
-
-                ret.add(step);
-            }
-        }
-
-        for (int i = 0, len = aabbs.size(); i < len; ++i) {
-            final AABB shape = aabbs.get(i);
-
-            final float step1 = (float)(shape.minY - box.minY);
-            final float step2 = (float)(shape.maxY - box.minY);
-
-            if (!(step1 < 0.0f) && step1 != collidedY && !(step1 > stepHeight)) {
-                ret.add(step1);
-            }
-
-            if (!(step2 < 0.0f) && step2 != collidedY && !(step2 > stepHeight)) {
-                ret.add(step2);
-            }
-        }
-
-        final float[] steps = ret.toFloatArray();
-        FloatArrays.unstableSort(steps);
-        return steps;
-    }
-    // Paper end - optimise collisions
-    // Paper start - optimise entity tracker
-    private net.minecraft.server.level.ChunkMap.TrackedEntity trackedEntity;
-
-    @Override
-    public final net.minecraft.server.level.ChunkMap.TrackedEntity moonrise$getTrackedEntity() {
-        return this.trackedEntity;
-    }
-
-    @Override
-    public final void moonrise$setTrackedEntity(final net.minecraft.server.level.ChunkMap.TrackedEntity trackedEntity) {
-        this.trackedEntity = trackedEntity;
-    }
-
-    private static void collectIndirectPassengers(final List<Entity> into, final List<Entity> from) {
-        for (final Entity passenger : from) {
-            into.add(passenger);
-            collectIndirectPassengers(into, ((Entity)(Object)passenger).passengers);
-        }
-    }
-    // Paper end - optimise entity tracker
 
     public Entity(EntityType<?> type, Level level) {
         this.type = type;
@@ -542,13 +427,6 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         this.position = Vec3.ZERO;
         this.blockPosition = BlockPos.ZERO;
         this.chunkPosition = ChunkPos.ZERO;
-        // Paper start - EAR 2
-        if (level != null) {
-            this.defaultActivationState = io.papermc.paper.entity.activation.ActivationRange.initializeEntityActivationState(this, level.spigotConfig);
-        } else {
-            this.defaultActivationState = false;
-        }
-        // Paper end - EAR 2
         SynchedEntityData.Builder builder = new SynchedEntityData.Builder(this);
         builder.define(DATA_SHARED_FLAGS_ID, (byte)0);
         builder.define(DATA_AIR_SUPPLY_ID, this.getMaxAirSupply());
@@ -1122,135 +1000,124 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             this.moveVector = movement;
         }
         try {
-        // Paper end - detailed watchdog information
-        if (this.noPhysics) {
-            this.setPos(this.getX() + movement.x, this.getY() + movement.y, this.getZ() + movement.z);
-            this.horizontalCollision = false;
-            this.verticalCollision = false;
-            this.verticalCollisionBelow = false;
-            this.minorHorizontalCollision = false;
-        } else {
-            if (type == MoverType.PISTON) {
-                // Paper start - EAR 2
-                this.activatedTick = Math.max(this.activatedTick, net.minecraft.server.MinecraftServer.currentTick + 20);
-                this.activatedImmunityTick = Math.max(this.activatedImmunityTick, net.minecraft.server.MinecraftServer.currentTick + 20);
-                // Paper end - EAR 2
-                movement = this.limitPistonMovement(movement);
-                if (movement.equals(Vec3.ZERO)) {
-                    return;
-                }
-            }
-
-            ProfilerFiller profilerFiller = Profiler.get();
-            profilerFiller.push("move");
-            if (this.stuckSpeedMultiplier.lengthSqr() > 1.0E-7) {
-                if (type != MoverType.PISTON) {
-                    movement = movement.multiply(this.stuckSpeedMultiplier);
-                }
-
-                this.stuckSpeedMultiplier = Vec3.ZERO;
-                this.setDeltaMovement(Vec3.ZERO);
-            }
-            // Paper start - ignore movement changes while inactive.
-            if (isTemporarilyActive && !(this instanceof ItemEntity) && movement == getDeltaMovement() && type == MoverType.SELF) {
-                setDeltaMovement(Vec3.ZERO);
-                profilerFiller.pop();
-                return;
-            }
-            // Paper end
-
-            movement = this.maybeBackOffFromEdge(movement, type);
-            Vec3 vec3 = this.collide(movement);
-            double d = vec3.lengthSqr();
-            if (d > 1.0E-7 || movement.lengthSqr() - d < 1.0E-7) {
-                if (this.fallDistance != 0.0 && d >= 1.0) {
-                    double min = Math.min(vec3.length(), 8.0);
-                    Vec3 vec31 = this.position().add(vec3.normalize().scale(min));
-                    BlockHitResult blockHitResult = this.level()
-                        .clip(new ClipContext(this.position(), vec31, ClipContext.Block.FALLDAMAGE_RESETTING, ClipContext.Fluid.WATER, this));
-                    if (blockHitResult.getType() != HitResult.Type.MISS) {
-                        this.resetFallDistance();
+            // Paper end - detailed watchdog information
+            if (this.noPhysics) {
+                this.setPos(this.getX() + movement.x, this.getY() + movement.y, this.getZ() + movement.z);
+                this.horizontalCollision = false;
+                this.verticalCollision = false;
+                this.verticalCollisionBelow = false;
+                this.minorHorizontalCollision = false;
+            } else {
+                if (type == MoverType.PISTON) {
+                    movement = this.limitPistonMovement(movement);
+                    if (movement.equals(Vec3.ZERO)) {
+                        return;
                     }
                 }
 
-                Vec3 vec32 = this.position();
-                Vec3 vec33 = vec32.add(vec3);
-                this.addMovementThisTick(new Entity.Movement(vec32, vec33, movement));
-                this.setPos(vec33);
-            }
+                ProfilerFiller profilerFiller = Profiler.get();
+                profilerFiller.push("move");
+                if (this.stuckSpeedMultiplier.lengthSqr() > 1.0E-7) {
+                    if (type != MoverType.PISTON) {
+                        movement = movement.multiply(this.stuckSpeedMultiplier);
+                    }
 
-            profilerFiller.pop();
-            profilerFiller.push("rest");
-            boolean flag = !Mth.equal(movement.x, vec3.x);
-            boolean flag1 = !Mth.equal(movement.z, vec3.z);
-            this.horizontalCollision = flag || flag1;
-            if (Math.abs(movement.y) > 0.0 || this.isLocalInstanceAuthoritative()) {
-                this.verticalCollision = movement.y != vec3.y;
-                this.verticalCollisionBelow = this.verticalCollision && movement.y < 0.0;
-                this.setOnGroundWithMovement(this.verticalCollisionBelow, this.horizontalCollision, vec3);
-            }
+                    this.stuckSpeedMultiplier = Vec3.ZERO;
+                    this.setDeltaMovement(Vec3.ZERO);
+                }
 
-            if (this.horizontalCollision) {
-                this.minorHorizontalCollision = this.isHorizontalCollisionMinor(vec3);
-            } else {
-                this.minorHorizontalCollision = false;
-            }
+                movement = this.maybeBackOffFromEdge(movement, type);
+                Vec3 vec3 = this.collide(movement);
+                double d = vec3.lengthSqr();
+                if (d > 1.0E-7 || movement.lengthSqr() - d < 1.0E-7) {
+                    if (this.fallDistance != 0.0 && d >= 1.0) {
+                        double min = Math.min(vec3.length(), 8.0);
+                        Vec3 vec31 = this.position().add(vec3.normalize().scale(min));
+                        BlockHitResult blockHitResult = this.level()
+                            .clip(new ClipContext(this.position(), vec31, ClipContext.Block.FALLDAMAGE_RESETTING, ClipContext.Fluid.WATER, this));
+                        if (blockHitResult.getType() != HitResult.Type.MISS) {
+                            this.resetFallDistance();
+                        }
+                    }
 
-            BlockPos onPosLegacy = this.getOnPosLegacy();
-            BlockState blockState = this.level().getBlockState(onPosLegacy);
-            if (this.isLocalInstanceAuthoritative()) {
-                this.checkFallDamage(vec3.y, this.onGround(), blockState, onPosLegacy);
-            }
+                    Vec3 vec32 = this.position();
+                    Vec3 vec33 = vec32.add(vec3);
+                    this.addMovementThisTick(new Entity.Movement(vec32, vec33, movement));
+                    this.setPos(vec33);
+                }
 
-            if (this.isRemoved()) {
                 profilerFiller.pop();
-            } else {
+                profilerFiller.push("rest");
+                boolean flag = !Mth.equal(movement.x, vec3.x);
+                boolean flag1 = !Mth.equal(movement.z, vec3.z);
+                this.horizontalCollision = flag || flag1;
+                if (Math.abs(movement.y) > 0.0 || this.isLocalInstanceAuthoritative()) {
+                    this.verticalCollision = movement.y != vec3.y;
+                    this.verticalCollisionBelow = this.verticalCollision && movement.y < 0.0;
+                    this.setOnGroundWithMovement(this.verticalCollisionBelow, this.horizontalCollision, vec3);
+                }
+
                 if (this.horizontalCollision) {
-                    Vec3 deltaMovement = this.getDeltaMovement();
-                    this.setDeltaMovement(flag ? 0.0 : deltaMovement.x, deltaMovement.y, flag1 ? 0.0 : deltaMovement.z);
+                    this.minorHorizontalCollision = this.isHorizontalCollisionMinor(vec3);
+                } else {
+                    this.minorHorizontalCollision = false;
                 }
 
-                if (this.canSimulateMovement()) {
-                    Block block = blockState.getBlock();
-                    if (movement.y != vec3.y) {
-                        block.updateEntityMovementAfterFallOn(this.level(), this);
-                    }
+                BlockPos onPosLegacy = this.getOnPosLegacy();
+                BlockState blockState = this.level().getBlockState(onPosLegacy);
+                if (this.isLocalInstanceAuthoritative()) {
+                    this.checkFallDamage(vec3.y, this.onGround(), blockState, onPosLegacy);
                 }
-                // CraftBukkit start
-                if (this.horizontalCollision && this.getBukkitEntity() instanceof org.bukkit.entity.Vehicle) {
-                    org.bukkit.entity.Vehicle vehicle = (org.bukkit.entity.Vehicle) this.getBukkitEntity();
-                    org.bukkit.block.Block block = this.level.getWorld().getBlockAt(Mth.floor(this.getX()), Mth.floor(this.getY()), Mth.floor(this.getZ()));
-
-                    if (movement.x > vec3.x) {
-                        block = block.getRelative(org.bukkit.block.BlockFace.EAST);
-                    } else if (movement.x < vec3.x) {
-                        block = block.getRelative(org.bukkit.block.BlockFace.WEST);
-                    } else if (movement.z > vec3.z) {
-                        block = block.getRelative(org.bukkit.block.BlockFace.SOUTH);
-                    } else if (movement.z < vec3.z) {
-                        block = block.getRelative(org.bukkit.block.BlockFace.NORTH);
+
+                if (this.isRemoved()) {
+                    profilerFiller.pop();
+                } else {
+                    if (this.horizontalCollision) {
+                        Vec3 deltaMovement = this.getDeltaMovement();
+                        this.setDeltaMovement(flag ? 0.0 : deltaMovement.x, deltaMovement.y, flag1 ? 0.0 : deltaMovement.z);
                     }
 
-                    if (!block.getType().isAir()) {
-                        org.bukkit.event.vehicle.VehicleBlockCollisionEvent event = new org.bukkit.event.vehicle.VehicleBlockCollisionEvent(vehicle, block, org.bukkit.craftbukkit.util.CraftVector.toBukkit(originalMovement)); // Paper - Expose pre-collision velocity
-                        event.callEvent();
+                    if (this.canSimulateMovement()) {
+                        Block block = blockState.getBlock();
+                        if (movement.y != vec3.y) {
+                            block.updateEntityMovementAfterFallOn(this.level(), this);
+                        }
                     }
-                }
-                // CraftBukkit end
+                    // CraftBukkit start
+                    if (this.horizontalCollision && this.getBukkitEntity() instanceof org.bukkit.entity.Vehicle) {
+                        org.bukkit.entity.Vehicle vehicle = (org.bukkit.entity.Vehicle) this.getBukkitEntity();
+                        org.bukkit.block.Block block = this.level.getWorld().getBlockAt(Mth.floor(this.getX()), Mth.floor(this.getY()), Mth.floor(this.getZ()));
+
+                        if (movement.x > vec3.x) {
+                            block = block.getRelative(org.bukkit.block.BlockFace.EAST);
+                        } else if (movement.x < vec3.x) {
+                            block = block.getRelative(org.bukkit.block.BlockFace.WEST);
+                        } else if (movement.z > vec3.z) {
+                            block = block.getRelative(org.bukkit.block.BlockFace.SOUTH);
+                        } else if (movement.z < vec3.z) {
+                            block = block.getRelative(org.bukkit.block.BlockFace.NORTH);
+                        }
 
-                if (!this.level().isClientSide() || this.isLocalInstanceAuthoritative()) {
-                    Entity.MovementEmission movementEmission = this.getMovementEmission();
-                    if (movementEmission.emitsAnything() && !this.isPassenger()) {
-                        this.applyMovementEmissionAndPlaySound(movementEmission, vec3, onPosLegacy, blockState);
+                        if (!block.getType().isAir()) {
+                            org.bukkit.event.vehicle.VehicleBlockCollisionEvent event = new org.bukkit.event.vehicle.VehicleBlockCollisionEvent(vehicle, block, org.bukkit.craftbukkit.util.CraftVector.toBukkit(originalMovement)); // Paper - Expose pre-collision velocity
+                            event.callEvent();
+                        }
                     }
-                }
+                    // CraftBukkit end
 
-                float blockSpeedFactor = this.getBlockSpeedFactor();
-                this.setDeltaMovement(this.getDeltaMovement().multiply(blockSpeedFactor, 1.0, blockSpeedFactor));
-                profilerFiller.pop();
+                    if (!this.level().isClientSide() || this.isLocalInstanceAuthoritative()) {
+                        Entity.MovementEmission movementEmission = this.getMovementEmission();
+                        if (movementEmission.emitsAnything() && !this.isPassenger()) {
+                            this.applyMovementEmissionAndPlaySound(movementEmission, vec3, onPosLegacy, blockState);
+                        }
+                    }
+
+                    float blockSpeedFactor = this.getBlockSpeedFactor();
+                    this.setDeltaMovement(this.getDeltaMovement().multiply(blockSpeedFactor, 1.0, blockSpeedFactor));
+                    profilerFiller.pop();
+                }
             }
-        }
-        // Paper start - detailed watchdog information
+            // Paper start - detailed watchdog information
         } finally {
             synchronized (this.posLock) { // Paper
                 this.moveVector = null;
@@ -1449,8 +1316,8 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             } else {
                 BlockState blockState = this.level().getBlockState(blockPos);
                 return (!(yOffset <= 0.5) || !blockState.is(BlockTags.FENCES))
-                        && !blockState.is(BlockTags.WALLS)
-                        && !(blockState.getBlock() instanceof FenceGateBlock)
+                    && !blockState.is(BlockTags.WALLS)
+                    && !(blockState.getBlock() instanceof FenceGateBlock)
                     ? blockPos.atY(Mth.floor(this.position.y - yOffset))
                     : blockPos;
             }
@@ -1522,144 +1389,242 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         return list.isEmpty() ? distance : -Shapes.collide(Direction.Axis.Y, boundingBox, list, -distance);
     }
 
-    // Paper start - optimise collisions
-    private Vec3 collide(Vec3 movement) {
-        final boolean xZero = movement.x == 0.0;
-        final boolean yZero = movement.y == 0.0;
-        final boolean zZero = movement.z == 0.0;
-        if (xZero & yZero & zZero) {
-            return movement;
-        }
+    //collision fix start
 
-        final AABB currentBox = this.getBoundingBox();
+    private Vec3 collide(Vec3 vec) {
+        AABB boundingBox = this.getBoundingBox();
+        List<VoxelShape> entityCollisions = this.level().getEntityCollisions(this, boundingBox.expandTowards(vec));
+        Vec3 vec3 = vec.lengthSqr() == 0.0 ? vec : collideBoundingBox(this, vec, boundingBox, this.level(), entityCollisions);
+        boolean flag = vec.x != vec3.x;
+        boolean flag1 = vec.y != vec3.y;
+        boolean flag2 = vec.z != vec3.z;
+        boolean flag3 = flag1 && vec.y < 0.0;
+        if (this.maxUpStep() > 0.0F && (flag3 || this.onGround()) && (flag || flag2)) {
+            AABB aabb = flag3 ? boundingBox.move(0.0, vec3.y, 0.0) : boundingBox;
+            AABB aabb1 = aabb.expandTowards(vec.x, this.maxUpStep(), vec.z);
+            if (!flag3) {
+                aabb1 = aabb1.expandTowards(0.0, -1.0E-5F, 0.0);
+            }
 
-        final List<VoxelShape> potentialCollisionsVoxel = new ArrayList<>();
-        final List<AABB> potentialCollisionsBB = new ArrayList<>();
+            List<VoxelShape> list = collectColliders(this, this.level, entityCollisions, aabb1);
+            float f = (float)vec3.y;
+            float[] floats = collectCandidateStepUpHeights(aabb, list, this.maxUpStep(), f);
 
-        final AABB initialCollisionBox;
-        if (xZero & zZero) {
-            // note: xZero & zZero -> collision on x/z == 0 -> no step height calculation
-            // this specifically optimises entities standing still
-            initialCollisionBox = movement.y < 0.0 ?
-                ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.cutDownwards(currentBox, movement.y) : ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.cutUpwards(currentBox, movement.y);
-        } else {
-            initialCollisionBox = currentBox.expandTowards(movement);
+            for (float f1 : floats) {
+                Vec3 vec31 = collideWithShapes(new Vec3(vec.x, f1, vec.z), aabb, list);
+                if (vec31.horizontalDistanceSqr() > vec3.horizontalDistanceSqr()) {
+                    double d = boundingBox.minY - aabb.minY;
+                    return vec31.subtract(0.0, d, 0.0);
+                }
+            }
         }
+        return vec3;
+    }
 
-        final List<AABB> entityAABBs = new ArrayList<>();
-        ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.getEntityHardCollisions(
-            this.level, (Entity)(Object)this, initialCollisionBox, entityAABBs, 0, null
-        );
-
-        ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.getCollisionsForBlocksOrWorldBorder(
-            this.level, (Entity)(Object)this, initialCollisionBox, potentialCollisionsVoxel, potentialCollisionsBB,
-            ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_FLAG_CHECK_BORDER, null
-        );
-        potentialCollisionsBB.addAll(entityAABBs);
-        final Vec3 collided = ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.performCollisions(movement, currentBox, potentialCollisionsVoxel, potentialCollisionsBB);
+    private static float[] collectCandidateStepUpHeights(AABB collisionBox, List<VoxelShape> collisions, float f, float stepHeight) {
+        FloatArraySet floatarrayset = new FloatArraySet(4);
+        java.util.Iterator iterator = collisions.iterator();
 
-        final boolean collidedX = collided.x != movement.x;
-        final boolean collidedY = collided.y != movement.y;
-        final boolean collidedZ = collided.z != movement.z;
+        while (iterator.hasNext()) {
+            VoxelShape voxelshape = (VoxelShape) iterator.next();
+            it.unimi.dsi.fastutil.doubles.DoubleList doublelist = voxelshape.getCoords(Direction.Axis.Y);
+            it.unimi.dsi.fastutil.doubles.DoubleListIterator doublelistiterator = doublelist.iterator();
 
-        final boolean collidedDownwards = collidedY && movement.y < 0.0;
+            while (doublelistiterator.hasNext()) {
+                double d0 = (Double) doublelistiterator.next();
+                float f2 = (float) (d0 - collisionBox.minY);
 
-        final double stepHeight;
+                if (f2 >= 0.0F && f2 != stepHeight) {
+                    if (f2 > f) {
+                        break;
+                    }
 
-        if ((!collidedDownwards && !this.onGround) || (!collidedX && !collidedZ) || (stepHeight = (double)this.maxUpStep()) <= 0.0) {
-            return collided;
+                    floatarrayset.add(f2);
+                }
+            }
         }
 
-        final AABB collidedYBox = collidedDownwards ? currentBox.move(0.0, collided.y, 0.0) : currentBox;
-        AABB stepRetrievalBox = collidedYBox.expandTowards(movement.x, stepHeight, movement.z);
-        if (!collidedDownwards) {
-            stepRetrievalBox = stepRetrievalBox.expandTowards(0.0, (double)-1.0E-5F, 0.0);
-        }
+        float[] afloat = floatarrayset.toFloatArray();
 
-        final List<VoxelShape> stepVoxels = new ArrayList<>();
-        final List<AABB> stepAABBs = entityAABBs;
+        FloatArrays.unstableSort(afloat);
+        return afloat;
+    }
 
-        ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.getCollisionsForBlocksOrWorldBorder(
-            this.level, (Entity)(Object)this, stepRetrievalBox, stepVoxels, stepAABBs,
-            ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.COLLISION_FLAG_CHECK_BORDER, null
-        );
+    private static Iterable<VoxelShape> concat(final Iterable<VoxelShape> a, final Iterable<VoxelShape> b) {
+        return () -> new java.util.Iterator<>() {
+            private java.util.Iterator<VoxelShape> ia = a.iterator();
+            private java.util.Iterator<VoxelShape> ib;
 
-        for (final float step : calculateStepHeights(collidedYBox, stepVoxels, stepAABBs, (float)stepHeight, (float)collided.y)) {
-            final Vec3 stepResult = ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.performCollisions(new Vec3(movement.x, (double)step, movement.z), collidedYBox, stepVoxels, stepAABBs);
-            if (stepResult.horizontalDistanceSqr() > collided.horizontalDistanceSqr()) {
-                return stepResult.add(0.0, collidedYBox.minY - currentBox.minY, 0.0);
+            @Override
+            public boolean hasNext() {
+                if (ia != null && ia.hasNext()) return true;
+                if (ia != null) ia = null;
+                if (ib == null) ib = b.iterator();
+                return ib.hasNext();
             }
+
+            @Override
+            public VoxelShape next() {
+                if (ia != null && ia.hasNext()) return ia.next();
+                if (ia != null) ia = null;
+                if (ib == null) ib = b.iterator();
+                return ib.next();
+            }
+        };
+    }
+
+    private static List<VoxelShape> collectEntityAndBorderColliders(
+        @Nullable Entity entity,
+        Level world,
+        List<VoxelShape> regularCollisions,
+        AABB movingEntityBoundingBox
+    ) {
+        com.google.common.collect.ImmutableList.Builder<VoxelShape> builder =
+            com.google.common.collect.ImmutableList.builderWithExpectedSize(regularCollisions.size() + 1);
+
+        if (!regularCollisions.isEmpty()) {
+            builder.addAll(regularCollisions);
+        }
+
+        WorldBorder border = world.getWorldBorder();
+        boolean addBorder = entity != null && border.isInsideCloseToBorder(entity, movingEntityBoundingBox);
+        if (addBorder) {
+            builder.add(border.getCollisionShape());
         }
 
-        return collided;
-        // Paper end - optimise collisions
+        return builder.build();
     }
 
-    private static float[] collectCandidateStepUpHeights(AABB box, List<VoxelShape> colliders, float deltaY, float maxUpStep) {
-        FloatSet set = new FloatArraySet(4);
+    private static Vec3 collideWithShapesAxisOnly(
+        @Nullable Entity entity,
+        Level world,
+        Vec3 movement,
+        AABB entityBoundingBox,
+        List<VoxelShape> entityAndBorder
+    ) {
+        double dx = movement.x;
+        double dy = movement.y;
+        double dz = movement.z;
+
+        if (dy != 0.0D) {
+            AABB yBox = entityBoundingBox.expandTowards(0.0D, dy, 0.0D);
+            Iterable<VoxelShape> yBlocks = world.getBlockCollisions(entity, yBox);
+            dy = Shapes.collide(Direction.Axis.Y, entityBoundingBox, concat(entityAndBorder, yBlocks), dy);
+            if (dy != 0.0D) {
+                entityBoundingBox = entityBoundingBox.move(0.0D, dy, 0.0D);
+            }
+        }
 
-        for (VoxelShape voxelShape : colliders) {
-            for (double d : voxelShape.getCoords(Direction.Axis.Y)) {
-                float f = (float)(d - box.minY);
-                if (!(f < 0.0F) && f != maxUpStep) {
-                    if (f > deltaY) {
-                        break;
-                    }
+        boolean swap = Math.abs(dx) < Math.abs(dz);
 
-                    set.add(f);
-                }
+        if (swap && dz != 0.0D) {
+            AABB zBox = entityBoundingBox.expandTowards(0.0D, 0.0D, dz);
+            Iterable<VoxelShape> zBlocks = world.getBlockCollisions(entity, zBox);
+            dz = Shapes.collide(Direction.Axis.Z, entityBoundingBox, concat(entityAndBorder, zBlocks), dz);
+            if (dz != 0.0D) {
+                entityBoundingBox = entityBoundingBox.move(0.0D, 0.0D, dz);
             }
         }
 
-        float[] floats = set.toFloatArray();
-        FloatArrays.unstableSort(floats);
-        return floats;
-    }
+        if (dx != 0.0D) {
+            AABB xBox = entityBoundingBox.expandTowards(dx, 0.0D, 0.0D);
+            Iterable<VoxelShape> xBlocks = world.getBlockCollisions(entity, xBox);
+            dx = Shapes.collide(Direction.Axis.X, entityBoundingBox, concat(entityAndBorder, xBlocks), dx);
+            if (!swap && dx != 0.0D) {
+                entityBoundingBox = entityBoundingBox.move(dx, 0.0D, 0.0D);
+            }
+        }
+
+        if (!swap && dz != 0.0D) {
+            AABB zBox = entityBoundingBox.expandTowards(0.0D, 0.0D, dz);
+            Iterable<VoxelShape> zBlocks = world.getBlockCollisions(entity, zBox);
+            dz = Shapes.collide(Direction.Axis.Z, entityBoundingBox, concat(entityAndBorder, zBlocks), dz);
+        }
 
-    public static Vec3 collideBoundingBox(@Nullable Entity entity, Vec3 vec, AABB collisionBox, Level level, List<VoxelShape> potentialHits) {
-        List<VoxelShape> list = collectColliders(entity, level, potentialHits, collisionBox.expandTowards(vec));
-        return collideWithShapes(vec, collisionBox, list);
+        return new Vec3(dx, dy, dz);
     }
 
-    public static List<VoxelShape> collectAllColliders(@Nullable Entity entity, Level level, AABB boundingBox) {
-        List<VoxelShape> entityCollisions = level.getEntityCollisions(entity, boundingBox);
-        return collectColliders(entity, level, entityCollisions, boundingBox);
+    public static Vec3 collideBoundingBox(
+        @Nullable Entity entity,
+        Vec3 movement,
+        AABB entityBoundingBox,
+        Level world,
+        List<VoxelShape> entityCollisions
+    ) {
+        if (movement.lengthSqr() == 0.0) {
+            return movement;
+        }
+
+        AABB expanded = entityBoundingBox.expandTowards(movement);
+        List<VoxelShape> entityAndBorder = collectEntityAndBorderColliders(entity, world, entityCollisions, expanded);
+
+        return collideWithShapesAxisOnly(entity, world, movement, entityBoundingBox, entityAndBorder);
     }
 
-    private static List<VoxelShape> collectColliders(@Nullable Entity entity, Level level, List<VoxelShape> collisions, AABB boundingBox) {
-        Builder<VoxelShape> builder = ImmutableList.builderWithExpectedSize(collisions.size() + 1);
-        if (!collisions.isEmpty()) {
-            builder.addAll(collisions);
+    private static List<VoxelShape> collectColliders(@Nullable Entity entity, Level world, List<VoxelShape> regularCollisions, AABB movingEntityBoundingBox) {
+        Builder<VoxelShape> builder = ImmutableList.builderWithExpectedSize(regularCollisions.size() + 1);
+
+        if (!regularCollisions.isEmpty()) {
+            builder.addAll(regularCollisions);
         }
 
-        WorldBorder worldBorder = level.getWorldBorder();
-        boolean flag = entity != null && worldBorder.isInsideCloseToBorder(entity, boundingBox);
+        WorldBorder worldborder = world.getWorldBorder();
+        boolean flag = entity != null && worldborder.isInsideCloseToBorder(entity, movingEntityBoundingBox);
+
         if (flag) {
-            builder.add(worldBorder.getCollisionShape());
+            builder.add(worldborder.getCollisionShape());
         }
 
-        builder.addAll(level.getBlockCollisions(entity, boundingBox));
+        builder.addAll(world.getBlockCollisions(entity, movingEntityBoundingBox));
         return builder.build();
     }
 
-    private static Vec3 collideWithShapes(Vec3 deltaMovement, AABB entityBB, List<VoxelShape> shapes) {
-        if (shapes.isEmpty()) {
-            return deltaMovement;
+    private static Vec3 collideWithShapes(Vec3 movement, AABB entityBoundingBox, List<VoxelShape> collisions) {
+        if (collisions.isEmpty()) {
+            return movement;
         } else {
-            Vec3 vec3 = Vec3.ZERO;
+            double d0 = movement.x;
+            double d1 = movement.y;
+            double d2 = movement.z;
+
+            if (d1 != 0.0D) {
+                d1 = Shapes.collide(Direction.Axis.Y, entityBoundingBox, collisions, d1);
+                if (d1 != 0.0D) {
+                    entityBoundingBox = entityBoundingBox.move(0.0D, d1, 0.0D);
+                }
+            }
+
+            boolean flag = Math.abs(d0) < Math.abs(d2);
+
+            if (flag && d2 != 0.0D) {
+                d2 = Shapes.collide(Direction.Axis.Z, entityBoundingBox, collisions, d2);
+                if (d2 != 0.0D) {
+                    entityBoundingBox = entityBoundingBox.move(0.0D, 0.0D, d2);
+                }
+            }
 
-            for (Direction.Axis axis : Direction.axisStepOrder(deltaMovement)) {
-                double d = deltaMovement.get(axis);
-                if (d != 0.0) {
-                    double d1 = Shapes.collide(axis, entityBB.move(vec3), shapes, d);
-                    vec3 = vec3.with(axis, d1);
+            if (d0 != 0.0D) {
+                d0 = Shapes.collide(Direction.Axis.X, entityBoundingBox, collisions, d0);
+                if (!flag && d0 != 0.0D) {
+                    entityBoundingBox = entityBoundingBox.move(d0, 0.0D, 0.0D);
                 }
             }
 
-            return vec3;
+            if (!flag && d2 != 0.0D) {
+                d2 = Shapes.collide(Direction.Axis.Z, entityBoundingBox, collisions, d2);
+            }
+
+            return new Vec3(d0, d1, d2);
         }
     }
 
+    //collision fix end
+
+    public static List<VoxelShape> collectAllColliders(@Nullable Entity entity, Level level, AABB boundingBox) {
+        List<VoxelShape> entityCollisions = level.getEntityCollisions(entity, boundingBox);
+        return collectColliders(entity, level, entityCollisions, boundingBox);
+    }
+
     protected float nextStep() {
         return (int)this.moveDist + 1;
     }
@@ -2529,11 +2494,11 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         // CraftBukkit end
         try {
             if (includeAll) { // CraftBukkit - selectively save position
-            if (this.vehicle != null) {
-                output.store("Pos", Vec3.CODEC, new Vec3(this.vehicle.getX(), this.getY(), this.vehicle.getZ()));
-            } else {
-                output.store("Pos", Vec3.CODEC, this.position());
-            }
+                if (this.vehicle != null) {
+                    output.store("Pos", Vec3.CODEC, new Vec3(this.vehicle.getX(), this.getY(), this.vehicle.getZ()));
+                } else {
+                    output.store("Pos", Vec3.CODEC, this.position());
+                }
             } // CraftBukkit
 
             output.store("Motion", Vec3.CODEC, this.getDeltaMovement());
@@ -2556,10 +2521,10 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             output.putInt("PortalCooldown", this.portalCooldown);
             // CraftBukkit start - selectively save uuid and world
             if (includeAll) {
-            output.store("UUID", UUIDUtil.CODEC, this.getUUID());
-            // PAIL: Check above UUID reads 1.8 properly, ie: UUIDMost / UUIDLeast
-            output.putLong("WorldUUIDLeast", this.level.getWorld().getUID().getLeastSignificantBits());
-            output.putLong("WorldUUIDMost", this.level.getWorld().getUID().getMostSignificantBits());
+                output.store("UUID", UUIDUtil.CODEC, this.getUUID());
+                // PAIL: Check above UUID reads 1.8 properly, ie: UUIDMost / UUIDLeast
+                output.putLong("WorldUUIDLeast", this.level.getWorld().getUID().getLeastSignificantBits());
+                output.putLong("WorldUUIDMost", this.level.getWorld().getUID().getMostSignificantBits());
             }
             output.putInt("Bukkit.updateLevel", Entity.CURRENT_LEVEL);
             if (!this.persist) {
@@ -2887,110 +2852,21 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     public boolean isInWall() {
-        // Paper start - optimise collisions
         if (this.noPhysics) {
             return false;
-        }
-
-        final double reducedWith = (double)(this.dimensions.width() * 0.8F);
-        final AABB boundingBox = AABB.ofSize(this.getEyePosition(), reducedWith, 1.0E-6D, reducedWith);
-        final Level world = this.level;
-
-        if (ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.isEmpty(boundingBox)) {
-            return false;
-        }
-
-        final int minBlockX = Mth.floor(boundingBox.minX);
-        final int minBlockY = Mth.floor(boundingBox.minY);
-        final int minBlockZ = Mth.floor(boundingBox.minZ);
-
-        final int maxBlockX = Mth.floor(boundingBox.maxX);
-        final int maxBlockY = Mth.floor(boundingBox.maxY);
-        final int maxBlockZ = Mth.floor(boundingBox.maxZ);
-
-        final int minChunkX = minBlockX >> 4;
-        final int minChunkY = minBlockY >> 4;
-        final int minChunkZ = minBlockZ >> 4;
-
-        final int maxChunkX = maxBlockX >> 4;
-        final int maxChunkY = maxBlockY >> 4;
-        final int maxChunkZ = maxBlockZ >> 4;
-
-        final int minSection = ca.spottedleaf.moonrise.common.util.WorldUtil.getMinSection(world);
-        final net.minecraft.world.level.chunk.ChunkSource chunkSource = world.getChunkSource();
-        final BlockPos.MutableBlockPos mutablePos = new BlockPos.MutableBlockPos();
-
-        for (int currChunkZ = minChunkZ; currChunkZ <= maxChunkZ; ++currChunkZ) {
-            for (int currChunkX = minChunkX; currChunkX <= maxChunkX; ++currChunkX) {
-                final net.minecraft.world.level.chunk.LevelChunkSection[] sections = chunkSource.getChunk(currChunkX, currChunkZ, net.minecraft.world.level.chunk.status.ChunkStatus.FULL, true).getSections();
-
-                for (int currChunkY = minChunkY; currChunkY <= maxChunkY; ++currChunkY) {
-                    final int sectionIdx = currChunkY - minSection;
-                    if (sectionIdx < 0 || sectionIdx >= sections.length) {
-                        continue;
-                    }
-                    final net.minecraft.world.level.chunk.LevelChunkSection section = sections[sectionIdx];
-                    if (section.hasOnlyAir()) {
-                        // empty
-                        continue;
-                    }
-
-                    final net.minecraft.world.level.chunk.PalettedContainer<net.minecraft.world.level.block.state.BlockState> blocks = section.states;
-
-                    final int minXIterate = currChunkX == minChunkX ? (minBlockX & 15) : 0;
-                    final int maxXIterate = currChunkX == maxChunkX ? (maxBlockX & 15) : 15;
-                    final int minZIterate = currChunkZ == minChunkZ ? (minBlockZ & 15) : 0;
-                    final int maxZIterate = currChunkZ == maxChunkZ ? (maxBlockZ & 15) : 15;
-                    final int minYIterate = currChunkY == minChunkY ? (minBlockY & 15) : 0;
-                    final int maxYIterate = currChunkY == maxChunkY ? (maxBlockY & 15) : 15;
-
-                    for (int currY = minYIterate; currY <= maxYIterate; ++currY) {
-                        final int blockY = currY | (currChunkY << 4);
-                        mutablePos.setY(blockY);
-                        for (int currZ = minZIterate; currZ <= maxZIterate; ++currZ) {
-                            final int blockZ = currZ | (currChunkZ << 4);
-                            mutablePos.setZ(blockZ);
-                            for (int currX = minXIterate; currX <= maxXIterate; ++currX) {
-                                final int blockX = currX | (currChunkX << 4);
-                                mutablePos.setX(blockX);
-
-                                final BlockState blockState = blocks.get((currX) | (currZ << 4) | ((currY) << 8));
-
-                                if (((ca.spottedleaf.moonrise.patches.collisions.block.CollisionBlockState)blockState).moonrise$emptyCollisionShape()
-                                    || !blockState.isSuffocating(world, mutablePos)) {
-                                    continue;
-                                }
-
-                                // Yes, it does not use the Entity context stuff.
-                                final VoxelShape collisionShape = blockState.getCollisionShape(world, mutablePos);
-
-                                if (collisionShape.isEmpty()) {
-                                    continue;
-                                }
-
-                                final AABB toCollide = boundingBox.move(-(double)blockX, -(double)blockY, -(double)blockZ);
-
-                                final AABB singleAABB = ((ca.spottedleaf.moonrise.patches.collisions.shape.CollisionVoxelShape)collisionShape).moonrise$getSingleAABBRepresentation();
-                                if (singleAABB != null) {
-                                    if (ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.voxelShapeIntersect(singleAABB, toCollide)) {
-                                        return true;
-                                    }
-                                    continue;
-                                }
-
-                                if (ca.spottedleaf.moonrise.patches.collisions.CollisionUtil.voxelShapeIntersectNoEmpty(collisionShape, toCollide)) {
-                                    return true;
-                                }
-                                continue;
-                            }
-                        }
+        } else {
+            float f = this.dimensions.width() * 0.8F;
+            AABB aabb = AABB.ofSize(this.getEyePosition(), f, 1.0E-6, f);
+            return BlockPos.betweenClosedStream(aabb)
+                .anyMatch(
+                    pos -> {
+                        BlockState blockState = this.level().getBlockState(pos);
+                        return !blockState.isAir()
+                            && blockState.isSuffocating(this.level(), pos)
+                            && Shapes.joinIsNotEmpty(blockState.getCollisionShape(this.level(), pos).move(pos), Shapes.create(aabb), BooleanOp.AND);
                     }
-                }
-            }
+                );
         }
-
-        return false;
-        // Paper end - optimise collisions
     }
 
     public InteractionResult interact(Player player, InteractionHand hand) {
@@ -3094,11 +2970,11 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     public boolean shearOffAllLeashConnections(@Nullable Player player) {
-    // Paper start - EntityUnleashEvent - overload
+        // Paper start - EntityUnleashEvent - overload
         return this.shearOffAllLeashConnections(player, null);
     }
     public boolean shearOffAllLeashConnections(@Nullable Player player, @Nullable InteractionHand interactionHand) {
-    // Paper end - EntityUnleashEvent - overload
+        // Paper end - EntityUnleashEvent - overload
         boolean flag = this.dropAllLeashConnections(player, interactionHand); // Paper - EntityUnleashEvent - overload
         if (flag && this.level() instanceof ServerLevel serverLevel) {
             serverLevel.playSound(null, this.blockPosition(), SoundEvents.SHEARS_SNIP, player != null ? player.getSoundSource() : this.getSoundSource());
@@ -3108,11 +2984,11 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     public boolean dropAllLeashConnections(@Nullable Player player) {
-    // Paper start - EntityUnleashEvent - overload
+        // Paper start - EntityUnleashEvent - overload
         return dropAllLeashConnections(player, null);
     }
     public boolean dropAllLeashConnections(@Nullable Player player, @Nullable InteractionHand interactionHand) {
-    // Paper start - EntityUnleashEvent - overload
+        // Paper start - EntityUnleashEvent - overload
         List<Leashable> list = Leashable.leashableLeashedTo(this);
         boolean flag = false; // Paper - EntityUnleashEvent - compute flag later, events might prevent unleashing all connected leashables.
         if (this instanceof Leashable leashable && leashable.isLeashed()) {
@@ -3365,8 +3241,8 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             Entity orig = craft == null ? null : craft.getHandle();
             if (this.getBukkitEntity() instanceof org.bukkit.entity.Vehicle && passenger.getBukkitEntity() instanceof org.bukkit.entity.LivingEntity) {
                 org.bukkit.event.vehicle.VehicleExitEvent event = new org.bukkit.event.vehicle.VehicleExitEvent(
-                        (org.bukkit.entity.Vehicle) this.getBukkitEntity(),
-                        (org.bukkit.entity.LivingEntity) passenger.getBukkitEntity(), !suppressCancellation // Paper - Force entity dismount during teleportation
+                    (org.bukkit.entity.Vehicle) this.getBukkitEntity(),
+                    (org.bukkit.entity.LivingEntity) passenger.getBukkitEntity(), !suppressCancellation // Paper - Force entity dismount during teleportation
                 );
                 // Suppress during worldgen
                 if (this.valid) {
@@ -3497,8 +3373,8 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
                     if (portalDestination != null) {
                         ServerLevel level = portalDestination.newLevel();
                         if (serverLevel.isAllowedToEnterPortal(level) && (level.dimension() == serverLevel.dimension() || this.canTeleport(serverLevel, level))
-                            )
-                         {
+                        )
+                        {
                             this.teleport(portalDestination);
                         }
                     }
@@ -3746,10 +3622,10 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             pluginManager.callEvent(entityCombustEvent);
             if (!entityCombustEvent.isCancelled()) {
                 this.igniteForSeconds(entityCombustEvent.getDuration(), false);
-            // Paper start - fix EntityCombustEvent cancellation
+                // Paper start - fix EntityCombustEvent cancellation
             } else {
                 this.setRemainingFireTicks(this.remainingFireTicks - 1);
-            // Paper end - fix EntityCombustEvent cancellation
+                // Paper end - fix EntityCombustEvent cancellation
             }
             // CraftBukkit end
         }
@@ -3926,32 +3802,32 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         String string = this.level() == null ? "~NULL~" : this.level().toString();
         return this.removalReason != null
             ? String.format(
-                Locale.ROOT,
-                "%s['%s'/%d, uuid='%s', l='%s', x=%.2f, y=%.2f, z=%.2f, cpos=%s, tl=%d, v=%b, removed=%s]", // Paper - add more info
-                this.getClass().getSimpleName(),
-                this.getPlainTextName(),
-                this.id,
-                this.uuid, // Paper - add more info
-                string,
-                this.getX(),
-                this.getY(),
-                this.getZ(),
-                this.chunkPosition(), this.tickCount, this.valid, // Paper - add more info
-                this.removalReason
-            )
+            Locale.ROOT,
+            "%s['%s'/%d, uuid='%s', l='%s', x=%.2f, y=%.2f, z=%.2f, cpos=%s, tl=%d, v=%b, removed=%s]", // Paper - add more info
+            this.getClass().getSimpleName(),
+            this.getPlainTextName(),
+            this.id,
+            this.uuid, // Paper - add more info
+            string,
+            this.getX(),
+            this.getY(),
+            this.getZ(),
+            this.chunkPosition(), this.tickCount, this.valid, // Paper - add more info
+            this.removalReason
+        )
             : String.format(
-                Locale.ROOT,
-                "%s['%s'/%d, uuid='%s', l='%s', x=%.2f, y=%.2f, z=%.2f, cpos=%s, tl=%d, v=%b]", // Paper - add more info
-                this.getClass().getSimpleName(),
-                this.getPlainTextName(),
-                this.id,
-                this.uuid, // Paper - add more info
-                string,
-                this.getX(),
-                this.getY(),
-                this.getZ(),
-                this.chunkPosition(), this.tickCount, this.valid // Paper - add more info
-            );
+            Locale.ROOT,
+            "%s['%s'/%d, uuid='%s', l='%s', x=%.2f, y=%.2f, z=%.2f, cpos=%s, tl=%d, v=%b]", // Paper - add more info
+            this.getClass().getSimpleName(),
+            this.getPlainTextName(),
+            this.id,
+            this.uuid, // Paper - add more info
+            string,
+            this.getX(),
+            this.getY(),
+            this.getZ(),
+            this.chunkPosition(), this.tickCount, this.valid // Paper - add more info
+        );
     }
 
     public final boolean isInvulnerableToBase(DamageSource damageSource) {
@@ -4454,11 +4330,11 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     public final void setBoundingBox(AABB bb) {
         // CraftBukkit start - block invalid bounding boxes
         double minX = bb.minX,
-                minY = bb.minY,
-                minZ = bb.minZ,
-                maxX = bb.maxX,
-                maxY = bb.maxY,
-                maxZ = bb.maxZ;
+            minY = bb.minY,
+            minZ = bb.minZ,
+            maxX = bb.maxX,
+            maxY = bb.maxY,
+            maxZ = bb.maxZ;
         double len = bb.maxX - bb.minX;
         if (len < 0) maxX = minX;
         if (len > 64) maxX = minX + 64.0;
@@ -4580,17 +4456,15 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     public Iterable<Entity> getIndirectPassengers() {
-        // Paper start - optimise entity tracker
-        final List<Entity> ret = new ArrayList<>();
-
-        if (this.passengers.isEmpty()) {
-            return ret;
+        // Paper start - Optimize indirect passenger iteration
+        if (this.passengers.isEmpty()) { return ImmutableList.of(); }
+        ImmutableList.Builder<Entity> indirectPassengers = ImmutableList.builder();
+        for (Entity passenger : this.passengers) {
+            indirectPassengers.add(passenger);
+            indirectPassengers.addAll(passenger.getIndirectPassengers());
         }
-
-        collectIndirectPassengers(ret, this.passengers);
-
-        return ret;
-        // Paper end - optimise entity tracker
+        return indirectPassengers.build();
+        // Paper end - Optimize indirect passenger iteration
     }
 
     public int countPlayerPassengers() {
@@ -4739,126 +4613,77 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         return Mth.lerp(partialTick, this.yRotO, this.yRot);
     }
 
-    // Paper start - optimise collisions
-    public boolean updateFluidHeightAndDoFluidPushing(final TagKey<Fluid> fluid, final double flowScale) {
+    public boolean updateFluidHeightAndDoFluidPushing(TagKey<Fluid> fluidTag, double motionScale) {
         if (this.touchingUnloadedChunk()) {
             return false;
-        }
-
-        final AABB boundingBox = this.getBoundingBox().deflate(1.0E-3);
-
-        final Level world = this.level;
-        final int minSection = ca.spottedleaf.moonrise.common.util.WorldUtil.getMinSection(world);
-
-        final int minBlockX = Mth.floor(boundingBox.minX);
-        final int minBlockY = Math.max((minSection << 4), Mth.floor(boundingBox.minY));
-        final int minBlockZ = Mth.floor(boundingBox.minZ);
-
-        // note: bounds are exclusive in Vanilla, so we subtract 1 - our loop expects bounds to be inclusive
-        final int maxBlockX = Mth.ceil(boundingBox.maxX) - 1;
-        final int maxBlockY = Math.min((ca.spottedleaf.moonrise.common.util.WorldUtil.getMaxSection(world) << 4) | 15, Mth.ceil(boundingBox.maxY) - 1);
-        final int maxBlockZ = Mth.ceil(boundingBox.maxZ) - 1;
-
-        final boolean isPushable = this.isPushedByFluid();
-        final BlockPos.MutableBlockPos mutablePos = new BlockPos.MutableBlockPos();
-
-        Vec3 pushVector = Vec3.ZERO;
-        double totalPushes = 0.0;
-        double maxHeightDiff = 0.0;
-        boolean inFluid = false;
-
-        final int minChunkX = minBlockX >> 4;
-        final int maxChunkX = maxBlockX >> 4;
-
-        final int minChunkZ = minBlockZ >> 4;
-        final int maxChunkZ = maxBlockZ >> 4;
-
-        final net.minecraft.world.level.chunk.ChunkSource chunkSource = world.getChunkSource();
-
-        final int chunkLenX = maxChunkX - minChunkX + 1;
-        // chunk index = (x - minX) + (maxX-minX+1)*(z - minZ)
-        //             = x + (maxX-minX+1)*z - (minX + (maxX-minX+1)*minZ)
-        final int chunkOffset = -(minChunkX + chunkLenX*minChunkZ);
-        //             = x + (maxX-minX+1)*z + chunkOffset
-        final net.minecraft.world.level.chunk.LevelChunkSection[][] sections = new net.minecraft.world.level.chunk.LevelChunkSection[chunkLenX * (maxChunkZ - minChunkZ + 1)][];
-
-        // init chunks
-        for (int currChunkZ = minChunkZ; currChunkZ <= maxChunkZ; ++currChunkZ) {
-            for (int currChunkX = minChunkX; currChunkX <= maxChunkX; ++currChunkX) {
-                sections[currChunkX + chunkLenX*currChunkZ + chunkOffset] = chunkSource.getChunk(currChunkX, currChunkZ, net.minecraft.world.level.chunk.status.ChunkStatus.FULL, false).getSections();
-            }
-        }
-
-        for (int currX = minBlockX; currX <= maxBlockX; ++currX) {
-            for (int currY = minBlockY; currY <= maxBlockY; ++currY) {
-                for (int currZ = minBlockZ; currZ <= maxBlockZ; ++currZ) {
-                    final FluidState fluidState = sections[(currX >> 4) + chunkLenX*(currZ >> 4) + chunkOffset][(currY >> 4) - minSection]
-                        .states.get((currX & 15) | ((currZ & 15) << 4) | ((currY & 15) << 8)).getFluidState();
-
-                    if (fluidState.isEmpty() || !fluidState.is(fluid)) {
-                        continue;
-                    }
-
-                    mutablePos.set(currX, currY, currZ);
-
-                    // CraftBukkit start - store last lava contact location
-                    if (fluid == FluidTags.LAVA) {
-                        this.lastLavaContact = mutablePos.immutable();
-                    }
-                    // CraftBukkit end
-
-                    final double height = (double)((float)currY + fluidState.getHeight(world, mutablePos));
-                    final double diff = height - boundingBox.minY;
-
-                    if (diff < 0.0) {
-                        continue;
-                    }
-
-                    inFluid = true;
-                    maxHeightDiff = Math.max(maxHeightDiff, diff);
-
-                    if (!isPushable) {
-                        continue;
-                    }
-
-                    ++totalPushes;
-
-                    final Vec3 flow = fluidState.getFlow(world, mutablePos);
+        } else {
+            AABB aabb = this.getBoundingBox().deflate(0.001);
+            int floor = Mth.floor(aabb.minX);
+            int ceil = Mth.ceil(aabb.maxX);
+            int floor1 = Mth.floor(aabb.minY);
+            int ceil1 = Mth.ceil(aabb.maxY);
+            int floor2 = Mth.floor(aabb.minZ);
+            int ceil2 = Mth.ceil(aabb.maxZ);
+            double d = 0.0;
+            boolean isPushedByFluid = this.isPushedByFluid();
+            boolean flag = false;
+            Vec3 vec3 = Vec3.ZERO;
+            int i = 0;
+            BlockPos.MutableBlockPos mutableBlockPos = new BlockPos.MutableBlockPos();
+
+            for (int i1 = floor; i1 < ceil; i1++) {
+                for (int i2 = floor1; i2 < ceil1; i2++) {
+                    for (int i3 = floor2; i3 < ceil2; i3++) {
+                        mutableBlockPos.set(i1, i2, i3);
+                        FluidState fluidState = this.level().getFluidState(mutableBlockPos);
+                        if (fluidState.is(fluidTag)) {
+                            double d1 = i2 + fluidState.getHeight(this.level(), mutableBlockPos);
+                            if (d1 >= aabb.minY) {
+                                flag = true;
+                                d = Math.max(d1 - aabb.minY, d);
+                                if (isPushedByFluid) {
+                                    Vec3 flow = fluidState.getFlow(this.level(), mutableBlockPos);
+                                    if (d < 0.4) {
+                                        flow = flow.scale(d);
+                                    }
 
-                    if (maxHeightDiff < 0.4) {
-                        pushVector = pushVector.add(flow.scale(maxHeightDiff));
-                    } else {
-                        pushVector = pushVector.add(flow);
+                                    vec3 = vec3.add(flow);
+                                    i++;
+                                }
+                                // CraftBukkit start - store last lava contact location
+                                if (fluidTag == FluidTags.LAVA) {
+                                    this.lastLavaContact = mutableBlockPos.immutable();
+                                }
+                                // CraftBukkit end
+                            }
+                        }
                     }
                 }
             }
-        }
 
-        this.fluidHeight.put(fluid, maxHeightDiff);
+            if (vec3.length() > 0.0) {
+                if (i > 0) {
+                    vec3 = vec3.scale(1.0 / i);
+                }
 
-        if (pushVector == Vec3.ZERO) {
-            return inFluid;
-        }
+                if (!(this instanceof Player)) {
+                    vec3 = vec3.normalize();
+                }
 
-        // note: totalPushes != 0 as pushVector was changed
-        pushVector = pushVector.scale(1.0 / totalPushes);
-        final Vec3 currMovement = this.getDeltaMovement();
+                Vec3 deltaMovement = this.getDeltaMovement();
+                vec3 = vec3.scale(motionScale);
+                double d2 = 0.003;
+                if (Math.abs(deltaMovement.x) < 0.003 && Math.abs(deltaMovement.z) < 0.003 && vec3.length() < 0.0045000000000000005) {
+                    vec3 = vec3.normalize().scale(0.0045000000000000005);
+                }
 
-        if (!((Entity)(Object)this instanceof Player)) {
-            pushVector = pushVector.normalize();
-        }
+                this.setDeltaMovement(this.getDeltaMovement().add(vec3));
+            }
 
-        pushVector = pushVector.scale(flowScale);
-        if (Math.abs(currMovement.x) < 0.003 && Math.abs(currMovement.z) < 0.003 && pushVector.length() < 0.0045000000000000005) {
-            pushVector = pushVector.normalize().scale(0.0045000000000000005);
+            this.fluidHeight.put(fluidTag, d);
+            return flag;
         }
-
-        this.setDeltaMovement(currMovement.add(pushVector));
-
-        // note: inFluid = true here as pushVector != 0
-        return true;
     }
-    // Paper end - optimise collisions
 
     public boolean touchingUnloadedChunk() {
         AABB aabb = this.getBoundingBox().inflate(1.0);
@@ -4930,7 +4755,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     public void setDeltaMovement(Vec3 deltaMovement) {
         if (deltaMovement.isFinite()) {
             synchronized (this.posLock) { // Paper - detailed watchdog information
-            this.deltaMovement = deltaMovement;
+                this.deltaMovement = deltaMovement;
             } // Paper - detailed watchdog information
         }
     }
@@ -5018,22 +4843,13 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     }
 
     public final void setPosRaw(double x, double y, double z, boolean forceBoundingBoxUpdate) {
-        // Paper start - rewrite chunk system
-        if (this.updatingSectionStatus) {
-            LOGGER.error(
-                "Refusing to update position for entity " + this + " to position " + new Vec3(x, y, z)
-                    + " since it is processing a section status update", new Throwable()
-            );
-            return;
-        }
-        // Paper end - rewrite chunk system
         if (!checkPosition(this, x, y, z)) {
             return;
         }
         // Paper end - Block invalid positions and bounding box
         if (this.position.x != x || this.position.y != y || this.position.z != z) {
             synchronized (this.posLock) { // Paper - detailed watchdog information
-            this.position = new Vec3(x, y, z);
+                this.position = new Vec3(x, y, z);
             } // Paper - detailed watchdog information
             int floor = Mth.floor(x);
             int floor1 = Mth.floor(y);
@@ -5176,12 +4992,6 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
 
     @Override
     public final void setRemoved(Entity.RemovalReason removalReason, org.bukkit.event.entity.EntityRemoveEvent.@Nullable Cause cause) { // CraftBukkit - add Bukkit remove cause
-        // Paper start - rewrite chunk system
-        if (!((ca.spottedleaf.moonrise.patches.chunk_system.level.ChunkSystemLevel)this.level).moonrise$getEntityLookup().canRemoveEntity((Entity)(Object)this)) {
-            LOGGER.warn("Entity " + this + " is currently prevented from being removed from the world since it is processing section status updates", new Throwable());
-            return;
-        }
-        // Paper end - rewrite chunk system
         org.bukkit.craftbukkit.event.CraftEventFactory.callEntityRemoveEvent(this, cause); // CraftBukkit
         final boolean alreadyRemoved = this.removalReason != null; // Paper - Folia schedulers
         if (this.removalReason == null) {
@@ -5192,7 +5002,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
             this.stopRiding();
         }
 
-        if (this.removalReason != Entity.RemovalReason.UNLOADED_TO_CHUNK) { this.getPassengers().forEach(Entity::stopRiding); } // Paper - rewrite chunk system
+        this.getPassengers().forEach(Entity::stopRiding);
         this.levelCallback.onRemove(removalReason);
         this.onRemoval(removalReason);
         // Paper start - Folia schedulers
@@ -5216,11 +5026,6 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
         this.getBukkitEntity().taskScheduler.retire();
     }
     // Paper end - Folia schedulers
-    // Paper start - optimise Folia entity scheduler
-    public final void registerScheduler() {
-        this.getBukkitEntity().taskScheduler.registerTo(net.minecraft.server.MinecraftServer.getServer().entitySchedulerTickList);
-    }
-    // Paper end - optimise Folia entity scheduler
 
     @Override
     public void setLevelCallback(EntityInLevelCallback levelCallback) {
@@ -5231,7 +5036,7 @@ public abstract class Entity implements SyncedDataHolder, DebugValueSource, Name
     public boolean shouldBeSaved() {
         return (this.removalReason == null || this.removalReason.shouldSave())
             && !this.isPassenger()
-            && (!this.isVehicle() || !((ca.spottedleaf.moonrise.patches.chunk_system.entity.ChunkSystemEntity)this).moonrise$hasAnyPlayerPassengers()); // Paper - rewrite chunk system
+            && (!this.isVehicle() || !this.hasExactlyOnePlayerPassenger());
     }
 
     @Override
